src/NServiceBus.NHibernate/TimeoutPersisters/IncorrectIndexDetector.cs

```diff
@@ -0,0 +1,164 @@
+        }
+
+        Dictionary<string, string> MergeProperties(IDictionary<string, string> properties)
+        {
+            var props = new Dictionary<string, string>(dialect.DefaultProperties);
+            foreach (var prop in properties)
+            {
+                props[prop.Key] = prop.Value;
+            }
+            return props;
+        }
+
+        public void LogWarningIfTimeoutEntityIndexIsIncorrect()
+        {
+            try
+            {
+                connectionHelper.Prepare();
+                var connection = connectionHelper.Connection;
+
+                var index = GetIndex(connection, timeoutEntityMapping, TimeoutEntityMap.EndpointIndexName);
+                if (index == null)
+                {
+                    Logger.Warn($"Could not find {TimeoutEntityMap.EndpointIndexName} index. This may cause significant performance degradation of message deferral. Consult NServiceBus NHibernate persistence documentation for details on how to create this index.");
+                    return;
+                }
+
+                var indexCorrect = index.ColumnsCount == 2
+                    && index.ColumnNameAt(1).Equals(nameof(TimeoutEntity.Endpoint), StringComparison.InvariantCultureIgnoreCase)
+                    && index.ColumnNameAt(2).Equals(nameof(TimeoutEntity.Time), StringComparison.InvariantCultureIgnoreCase);
+
+                Logger.Debug($"Detected {TimeoutEntityMap.EndpointIndexName} ({index.ColumnNameAt(1)}, {index.ColumnNameAt(2)})");
+
+                if (!indexCorrect)
+                {
+                    Logger.Warn($"The {TimeoutEntityMap.EndpointIndexName} index has incorrect column order. This may cause significant performance degradation of message deferral. Consult NServiceBus NHibernate persistence documentation for details on how to create this index.");
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.Warn($"Could not inspect {TimeoutEntityMap.EndpointIndexName} index definition.", e);
+            }
+            finally
+            {
+                try
+                {
+                    connectionHelper.Release();
+                }
+                catch (Exception e)
+                {
+                    Logger.Error("Error closing connection", e);
+                }
+            }
+        }
+
+        private IIndex GetIndex(DbConnection connection, PersistentClass entity, string indexName)
+        {
+
+
+            // Check if running on SQL Server.
+            if (typeof(MsSql2005Dialect).IsAssignableFrom(dialect.GetType()))
+            {
+                var restrictions = new string[5]
+                {
+                    entity.Table.Catalog,
+                    entity.Table.Schema,
+                    entity.Table.Name,
+                    indexName,
+                    null
+                };
+                return new SqlServerIndex(connection.GetSchema("IndexColumns", restrictions));
```

Comment:
Nitpicking. Since in this path we already know we are in "SQL Server" zone, could we here transform the `DataTable` to a `string[]` containing the names of the columns? Then we would not need `IIndex` at all.